extensions/hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/CsvInputFormat.java

```diff
@@ -54,54 +59,87 @@ public RecordReader<NullWritable, Object> createRecordReader(InputSplit split, T
-                        throw new JetException("Header row missing in " + split);
-                    }
-                    String[] header = (String[]) iterator.next();
                     List<String> fieldNames = new ArrayList<>();
                     String field;
                     for (int i = 0; (field = configuration.get(CSV_INPUT_FORMAT_FIELD_LIST_PREFIX + i)) != null; i++) {
                         fieldNames.add(field);
                     }
+
                     projection = (Function) createFieldProjection(header, fieldNames);
                 } else {
-                    iterator = new CsvMapper().readerFor(formatClazz)
-                                              .withoutFeatures(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
-                                              .with(CsvSchema.emptySchema().withHeader())
-                                              .readValues((InputStream) in);
+                    Builder builder = CsvSchema.builder();
+                    for (String column : header) {
+                        builder.addColumn(column);
+                    }
+                    objectReader = new CsvMapper().readerFor(formatClazz)
+                                                  .withoutFeatures(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+                                                  .with(builder.build());
                 }
             }
 
-            @Override
-            public boolean nextKeyValue() {
-                if (!iterator.hasNext()) {
-                    return false;
+            private String[] readHeader(FileSplit fileSplit, TaskAttemptContext context) throws IOException {
+                Configuration configuration = context.getConfiguration();
+                Path file = fileSplit.getPath();
+                FileSystem fs = file.getFileSystem(configuration);
+
+                try (InputStream in = fs.open(file)) {
+                    try (InputStream wrapped = wrap(in, configuration, file)) {
+                        ObjectReader headerReader = new CsvMapper()
+                                .enable(Feature.WRAP_AS_ARRAY)
+                                .readerFor(String[].class)
+                                .with(CsvSchema.emptySchema().withSkipFirstDataRow(false));
+
+                        try (MappingIterator<Object> iterator = headerReader.readValues(wrapped)) {
+                            if (!iterator.hasNext()) {
+                                throw new JetException("Header row missing in " + fileSplit);
+                            }
+                            return (String[]) iterator.next();
+                        }
+                    }
+                }
+
+            }
+
+            private InputStream wrap(InputStream in, Configuration configuration, Path file) throws IOException {
+                CompressionCodec codec = new CompressionCodecFactory(configuration).getCodec(file);
+                if (codec != null) {
```

Comment:
nit: `return codec != null ? codec.createInputStream(in) : in`

Reply:
Done in ec0def77